### PR TITLE
Expunge the concept of well-known symbols from the checker

### LINKED ISSUE #24622: keyof does not include well known symbols
**TypeScript Version:**  2.9.1

**Search Terms:** 
keyof well known symbol

**Code**

```ts
const s = Symbol("s");

interface I {
    n: number;
    [s]: string;
    [Symbol.iterator](): IterableIterator<[string, string]>;
}

type K = keyof I;
type T = I[K];
```

**Expected behavior:**
```ts
type K = keyof I; // type K = typeof s | "n" | typeof Symbol.iterator
type T = I[K]; // type T = string | number | () => IterableIterator<[string, string]>
```

**Actual behavior:**
```ts
type K = keyof I; // type K = typeof s | "n"
type T = I[K]; // type T = string | number
```

### PR DESCRIPTION
**Pull request by weswigham**
Interestingly, there is some fallout here in some odd places:
* First of all, some small changes needed to occur to replicate old behavior. Right now, a computed name of type `any` produces a `number` index signature. With this change, we make a `string` index signature instead (it's more broad and closer to correct).
* Additionally, late-bound members weren't issuing duplicate declaration errors. With this change, they should.
* There's also a fourslash test changes due to the lib in fourslash not-actually-having-symbols (I looked into fixing it to use the built lib, it requires changing a bunch of tests, don't wanna do that here), so the test relied on the old behavior of well-known symbols always "resolving" uniquely, regardless of weather or not they were _actually_ defined.
* Lastly, some emit changes - specifically, where previously we'd elide `Symbol.iterator` when it appeared as a property name (and assume that the expression had no sideffects), now we retain and emit it (and potentially cache it), as we generally do not assume property access expressions are safe to elide or copy.

Fixes #24622

An important implementation note: As we discussed in person, Inside the checker, if we see members of the global `SymbolConstructor` of type `symbol`, we also _assume_ you meant to say `unique symbol`. In this way, we retain compataibility with older libs or definition files (even as they update), which is how this can build (at all), given node is shimming `Symbol.iterator` as a `symbol`.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3531,9 +3532,11 @@ namespace ts {
                         context.enclosingDeclaration = undefined;
                         if (getCheckFlags(propertySymbol) & CheckFlags.Late) {
                             const decl = first(propertySymbol.declarations);
-                            const name = hasLateBindableName(decl) && resolveEntityName(decl.name.expression, SymbolFlags.Value);
-                            if (name && context.tracker.trackSymbol) {
-                                context.tracker.trackSymbol(name, saveEnclosingDeclaration, SymbolFlags.Value);
+                            if (hasLateBindableName(decl)) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: This is a copy of the primary change from #24668 - it's required to make the symbols get looked up correctly.

**Comment by weswigham** [comment]
Huh. These changes seem pretty innocuous, but the build of `run.js` (the tests) runs out of memory on node6 (and only on node6). During sourcemap emit no less, so very late in the process. @rbuckton can you think of anything offhand we'd be doing where the memory consumption of ndoe6 vs node8 would differ enough to crash the process?

**Comment by RyanCavanaugh** [comment]
@weswigham Ron's sourcemap rewrite should be going in soon so that will hopefully fix the OOM issues you're getting here

**Comment by andrewbranch** [comment]
Also fixes #31253

**Comment by ExE-Boss** [comment]
This&nbsp;needs some&nbsp;serious&nbsp;rebasing.

**Comment by ExE-Boss** [comment]
This will also fix <https://github.com/microsoft/TypeScript/issues/36468>

**Comment by KilianKilmister** [comment]
any update on this?

**Comment by weswigham** [comment]
@rbuckton do we have an appetite for looking into this again? It looks like a number of things are getting dupe'd to it. IIRC we held off on it long ago so `lib` references could be a thing in `node`, to potentially remove the need for the global `SymbolConstructor`-`symbol`-as-`unique symbol` workaround.

**Comment by justingrant** [comment]
FWIW, this PR would also probably fix https://github.com/immerjs/immer/issues/710.

**Comment by DanielRosenwasser** [comment]
I'd like to bring this back to our focus. I recently spoke to @benlesh about RxJS and how the library leans on `Symbol.observable` existing.

So just to start, imagine an augmentation like

```ts
interface SymbolConstructor {
    observable: symbol;
}
```

This is inspired by the existing declarations; however, the [`symbol-observable`](https://github.com/benlesh/symbol-observable) library re-exports `Symbol.observable`. If the declaration file contains code like

```ts
const symbolObservable: typeof Symbol.observable;
export default symbolObservable;
```

that appears to any other code as just `symbol`. 

Trying to switch over to a unique symbol like so

```ts
interface SymbolConstructor {
    observable: unique symbol;
//              ^^^^^^
}
```

won't necessarily work because other libraries have also defined it as `symbol`

```ts
interface SymbolConstructor {
    observable: symbol;
}

interface SymbolConstructor {
    observable: unique symbol;
//  ~~~~~~~~~~
// Subsequent property declarations must have the same type.  Property 'observable' must be of type 'unique symbol', but here has type 'symbol'.
}
```

**Comment by benlesh** [comment]
I would like to note that:

1. I'm sorry. haha
2. I inherited `symbol-observable` which was always written this way, while it was in wide use.
3. This stuff is hard.

**Comment by weswigham** [comment]
https://github.com/microsoft/TypeScript/pull/42543 Now supersedes this as an updated version for 2021 - large parts are the same (and the result certainly is), but some of the implementation is different.

**Pull request closed without merging**